### DISCUSSION TITLE
[@types/node] `ReadableBase` async iterator conflicts with [@types/readable-stream] `Readable`

### DISCUSSION THREAD

**Discussion by james-pre** [DefinitelyTyped/DefinitelyTyped#71307] [Issues with a @types package]
When using `readable-stream`, I was getting type errors due to conflicting `AsyncIterator` types.

Reproduction:
```ts
import type * as fs from 'node:fs';
import { Readable, Writable } from 'readable-stream';

// polyfill `fs.ReadStream`
export class ExampleReadStream extends Readable implements fs.ReadStream {
	close(callback = () => null): void {
		// ...
	}

	declare bytesRead: number;
	declare path: string | Buffer;
	declare pending: boolean;
}
```

Dependency graph:

<img src="https://github.com/user-attachments/assets/42f8beb1-08b0-455d-ac85-b4268cfe3a68" width="500px" />

The problem:

`_Readable.iterator` and `_Readable[Symbol.asyncIterator]` return an `AsyncIterator` and `ReadableBase.iterator` and `ReadableBase[Symbol.asyncIterator]` return a `NodeJS.AsyncIterator`, but `NodeJS.AsyncIterator` is not compatible with `AsyncIterator`.

I took a look at how all of these types work with each other:

<img src="https://github.com/user-attachments/assets/0bedad42-0963-4d8a-b81b-9fccd0a2e3ac" width="500px" />

Analyzing the async iterator types, I was surprised to find that `NodeJS.AsyncIterator` actually *extends* `AsyncIterator`, so I made a reproduction:

```ts
interface AsyncItWithNode {
	it: NodeJS.AsyncIterator<any>;
}

interface AsyncIt {
	it: AsyncIterator<any>;
}

declare class Broken implements AsyncItWithNode {
	it: AsyncIterableIterator<any>; // <-- error!
}

declare class Working implements AsyncIt {
	it: AsyncIterableIterator<any>;
}
```

But here, we get the error:
```
Property 'it' in type 'Broken' is not assignable to the same property in base type 'AsyncItWithNode'.
  Property '[Symbol.asyncDispose]' is missing in type 'AsyncIterableIterator<any>' but required in type 'AsyncIterator<any, any, any>'.ts(2416)
lib.esnext.disposable.d.ts(40, 5): '[Symbol.asyncDispose]' is declared here.
```

Finally, we appear to have found the root cause.

Since `AsyncIteratorObject` extends `AsyncDisposable`, `NodeJS.AsyncIterator` has a `[Symbol.asyncDispose]` property. Typescript's `AsyncIterator` is missing this property since it doesn't extend from `AsyncIteratorObject`.

[a comment by an automated account is not shown]

**Comment by Renegade334** [comment]
Returned iterators in @types/node now extend `[Async]IteratorObject` in TS 5.6+. The `[Async]IterableIterator` interfaces (which are still used by readable-stream) are not assignable to `[Async]IteratorObject` when the esnext libs are loaded, as these add additional methods to those interfaces, hence the error.

It's probably not safe to assume that readable-stream types will always be interassignable with @types/node stream types, but in any case, the discussion would be best targeted at readable-stream to see if it can be updated to use the new @types/node interfaces for returned iterators.

**Reply by james-pre** [reply]
I can create a PR to update readable-stream's types.
